File: Python/YUKI/다리를 지나는 트럭.py

def Up(a,List,weight):
    if a + sum(List) <= weight: ######  a는 다음 트럭 무게
        return True
    return False

def Append(a,List):
    List[0] = a
    return List

def Go(List,bridge_length,weight,i,count,truck_weights):
    for count in range(bridge_length-1,0,-1):
        List[count] = List[count-1]
    if (sum(List[1:]) + truck_weights[i+1]) <= weight:
        Append(truck_weights[i+1],List)
    elif (sum(List[1:]) + truck_weights[i+1]) > weight:
        List[0] = 0
    count += 1
    return List,count,i
    
def solution(bridge_length, weight, truck_weights):
    List = [0]*bridge_length
    truck_weights.append(0)
    i=0
    count = 0
    Append(a,List)
    while i <=(len(truck_weights)-1):
        a = truck_weights[i]  #a는 선택된 트럭
        Go(List,bridge_length,weight,i,count,truck_weights)



    
solution(5,10,[7,4,5,6])

File: Javascript/Programmers/Level1/다리를지나는트럭.js
function solution(bridge_length, weight, truck_weights) {
    var answer = 0;
    var first,truck,last;
    var trucks = [];
    var total_weight = 0;
    while(truck_weights.length ){
        first = truck_weights[0];
        if (trucks.length==bridge_length){
            last = trucks.pop()
            total_weight -= last
        }
        if (total_weight + first <= weight){
            first = truck_weights.shift();
            trucks.unshift(first);
            total_weight += first;
        } else if (total_weight + first > weight){
            trucks.unshift(0)
        }
        answer += 1
        
    } answer += bridge_length
    return answer;
}

var bridge_length, weight, truck_weights;
var problem = [
    [2,10,[7,4,5,6]],
    [100,100,[10]],
    [100,100,[10,10,10,10,10,10,10,10,10,10]]
]
for (i=0;i<problem.length; i++){
    bridge_length = problem[i][0];
    weight = problem[i][1];
    truck_weights = problem[i][2];
    alert(solution(bridge_length, weight, truck_weights));
}